### PR TITLE
Fix vectorized toLowerCase and toUppercase intrinsics on Z

### PR DESCRIPTION
**Pull request by fjeremic**
In #4597 we identified cases on which `toUpperCase` fails on Linux on 
Z. This is due to some faulty logic in the evaluator for these 
intrinsics on Z. Bugs seem to have been introduced in #2318 for certain 
characters. Namely the test in #4597 fails to convert the `\u00FF` 
character. The bug is obvious, since `toUpperCase` of such a character 
cannot be handled and the JIT evaluator determines the invalid 
character range by checking if the character is greater than 0xFF, but 
it should be checking if greater than or equal to.

In addition the compressed String evaluator also fails on 0xB5 as 
determined by the unit tests. Both issues are fixed.

Fixes: #4597

### CODE REVIEW

**Comment by fjeremic** [comment]
I've improved the tests by including an exhaustive list of tests for uppercasing and lowercasing. There were two issues:

1. The previous testing never tested all characters so some conversions were wrong and uncaught
2. The whole point of Test_JITHelpers is to compile the intrinsics and we should be testing with `-Xjit:count=0` as a variation

Both issues will be fixed. There will be a subsequent commit fixing the issues, but before I add it I'd like to kick off sanity testing with the new tests and ensure the failures are caught on the farm. In addition to the problem observed in #4597 we have one additional character we're not converting properly (0xB5). Hopefully both failures show up on the builds.

Jenkins test sanity xlinux,zlinux JDK8 JDK11

**Comment by fjeremic** [comment]
I've improved the tests by including an exhaustive list of tests for uppercasing and lowercasing. There were two issues:

1. The previous testing never tested all characters so some conversions were wrong and uncaught
2. The whole point of Test_JITHelpers is to compile the intrinsics and we should be testing with `-Xjit:count=0` as a variation

Both issues will be fixed. There will be a subsequent commit fixing the issues, but before I add it I'd like to kick off sanity testing with the new tests and ensure the failures are caught on the farm. In addition to the problem observed in #4597 we have one additional character we're not converting properly (0xB5). Hopefully both failures show up on the builds.

Jenkins test sanity xlinux,zlinux JDK8,JDK11

**Comment by fjeremic** [comment]
As expected the tests failed as follows [1]:

```
FAILED: test_toUpperIntrinsicLatin1
java.lang.AssertionError: Failed to correctly uppercase character 0xb5, actual = 0xb5 expected = 0x39c expected [true] but found [false]
	at org.testng.Assert.fail(Assert.java:96)
	at org.testng.Assert.failNotEquals(Assert.java:776)
	at org.testng.Assert.assertTrue(Assert.java:44)
	at org.openj9.test.com.ibm.jit.Test_JITHelpers.test_toUpperIntrinsicLatin1(Test_JITHelpers.java:297)
...
FAILED: test_toUpperIntrinsicUTF16
java.lang.AssertionError: Failed to correctly uppercase character 0xff, actual = 0xff expected = 0x178 expected [true] but found [false]
	at org.testng.Assert.fail(Assert.java:96)
	at org.testng.Assert.failNotEquals(Assert.java:776)
	at org.testng.Assert.assertTrue(Assert.java:44)
	at org.openj9.test.com.ibm.jit.Test_JITHelpers.test_toUpperIntrinsicUTF16(Test_JITHelpers.java:261)
```

The second failure is the one observed in #4597. I'll test my fix locally tomorrow morning and attach it as a new commit to this PR.

[1] https://ci.eclipse.org/openj9/job/PullRequest-Sanity-JDK11-linux_390-64_cmprssptrs-OpenJ9/198/tapResults/

**Comment by fjeremic** [comment]
And the fix is in. I've tested manually the tests pass but for good measure we'll test on the farm as well. Requesting reviews from one of @smlambert / @llxia for the test piece of the PR and one of @NigelYiboYu / @dhong44 for the Z codegen piece.

Jenkins test sanity xlinux,zlinux JDK8,JDK11

File: runtime/compiler/z/codegen/J9TreeEvaluator.cpp [hunk 1]
```diff
@@ -668,36 +669,36 @@ TR::Register * caseConversionHelper(TR::Node* node, TR::CodeGenerator* cg, bool
...
 
-   // Can't toUpper \u00df (capital sharp s) nor \u00b5 (mu) with a simple addition of 0x20
-   // Condition code equal for capital sharp and mu (bit 0=0x80) and greater than (bit 2=0x20) for codes larger than 0xFF
    if (isToUpper)
       {
+      // Can't uppercase \u00DF (capital sharp s) nor \u00B5 (mu) with a simple addition of 0x20 so we do an equality
+      // comparison (bit 0) and greater than or equal comparison (bits 0 and 2) for codes larger than or equal to 0xFF
+      generateVRIaInstruction(cg, TR::InstOpCode::VLEIH, node, invalidRangeVector, 0xDFDF, 0x0);
+      generateVRIaInstruction(cg, TR::InstOpCode::VLEIH, node, invalidRangeVector, 0xB5B5, 0x1);
+      generateVRIaInstruction(cg, TR::InstOpCode::VLEIH, node, invalidRangeVector, 0xFFFF, 0x2);
```
Comment (NigelYiboYu) [inline runtime/compiler/z/codegen/J9TreeEvaluator.cpp]: LGTM

Comment (dhong44) [reply runtime/compiler/z/codegen/J9TreeEvaluator.cpp]: LGTM

**Comment by fjeremic** [comment]
@andrewcraik @0dvictor FYI we'll need to take a look at these x86 failures as my unit tests have caught what are likely bugs in the x86 `toUpperCase` and `toLowerCase` intrinsics. The output seems correct however so perhaps something else went wrong on the x86 side.

**Comment by andrewcraik** [comment]
per discussion the issues on x86 are that we need to truncate the return value of the helper to be 0 or 1 per recent spec updates, ensure we check the `disableHWAcceleratedStringCaseConv` flag. With that the functional issues should be resolved and a remaining perf item will be to expand the number of characters converted by the Latin1 helpers which are currently limited to 0-127 (eg core ASCII). @0dvictor is having a look and coordinating with @fjeremic directly.

**Comment by fjeremic** [comment]
Jenkins test sanity xlinux,zlinux JDK8,JDK11

**Comment by fjeremic** [comment]
Jenkins test sanity xlinux,zlinux JDK8,JDK11

**Comment by fjeremic** [comment]
All issues have been fixed. This is ready for final review from committers.

**Comment by 0xdaryl** [comment]
@andrewcraik, can you give your +1 if you're fine with the x86 changes please?

**Pull request merged**